src/imperative/cached_op.cc

```diff
@@ -32,247 +32,12 @@ DMLC_REGISTER_PARAMETER(CachedOpConfig);
 
 constexpr uint32_t kEidNotExist = std::numeric_limits<uint32_t>::max();
 
-const char CachedOp::FULL[] = "full";
-const char CachedOp::FORWARD[] = "forward";
-const char CachedOp::BACKWARD[] = "backward";
-const char CachedOp::REF_COUNT[] = "ref_count";
-const char CachedOp::MEM_PLAN[] = "mem_plan";
-const char CachedOp::STORAGE_PLAN[] = "storage_plan";
-
-namespace {
-
-std::string AddPrefix(const std::string& prefix,
-                      const std::string& s) {
-  return prefix + "_" + s;
-}
-
-}  // namespace
-
-struct CachedOp::GraphInfo {
-  nnvm::Graph fwd_graph;
-  nnvm::Graph grad_graph;
-  nnvm::Graph full_graph;
-  std::vector<nnvm::NodeEntry> ograd_entries;
-  std::unordered_map<uint32_t, uint32_t> fwd_input_to_grad_output;
-  std::vector<OpReqType> bwd_output_reqs;
-  std::vector<uint32_t> bwd_input_eid;
-};
-
 struct CachedOp::DynamicRuntime {
   GraphInfo info;
   std::vector<NDArray> buff;
   std::vector<OpStatePtr> op_states;
 };
 
-void CreateFullGraph(const nnvm::Symbol& sym,
```

Comment:
Any particular reason you moved these function implementation from .cc to .h?

Reply:
to allow to be used in both CachedOp and CachedOpThreadSafe.